### DISCUSSION TITLE
Is it possible to reshape and reorder text using harfbuzz for use in unsupported applications?

### DISCUSSION THREAD

**Discussion by NightMachinery** [harfbuzz/harfbuzz#2903] [Q&A] [answered]
I have seen https://harfbuzz.github.io/utilities.html , but the output seems to always do the rendering part within harfbuzz itself; I just want normal text to come out, but reshaped and reordered. I have a simple CLI [tool](https://github.com/NightMachinary/rtl_reshaper_rs) that does this using `arabic_reshaper` and `unic-bidi`, but those packages are buggy. I also have a Python [script](https://github.com/NightMachinary/.shells/blob/master/scripts/python/RTL/rtl_reshaper.py) for doing this, but its underlying python libraries are extremely slow.

**Comment by khaledhosny** [comment, marked as answer]
HarfBuzz output is glyph indices and position, not strings, so it can’t be used where strings are expected. Your best option for cases like these is to export images with `hb-view` and insert them in the desired application, which works for simple, one-line strings.

**Comment by behdad** [comment]
If you only want Arabic+Hebrew, then FriBidi can do that for you.

**Comment by behdad** [comment]
> I have a simple CLI [tool](https://github.com/NightMachinary/rtl_reshaper_rs) that does this using `arabic_reshaper` and `unic-bidi`, but those packages are buggy. I also have a Python [script](https://github.com/NightMachinary/.shells/blob/master/scripts/python/RTL/rtl_reshaper.py) for doing this, but its underlying python libraries are extremely slow.

FriBidi can do exactly that.
